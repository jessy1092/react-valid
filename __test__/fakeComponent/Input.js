import React from 'react';

const Input = ({ validate, message }) => (
  <div>
    <input onChange={e => validate(e.target.value)} />
    <div>{message}</div>
  </div>
);

export default Input;
