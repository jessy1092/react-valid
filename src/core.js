import React from 'react';

export const addMethod = (validators, name, method = () => {}, message = '') => ({
  ...validators,
  [name]: {
    method,
    message,
  },
});

export const addMethods = (validators, appendValidators = {}) => {
  const defaultValidator = {
    method() {},
    message: '',
  };

  const newValidates = {
    ...validators,
  };

  Object.keys(appendValidators).forEach(name => {
    newValidates[name] = {
      ...defaultValidator, // Set default value to prevent method not define
      ...appendValidators[name],
    };
  });

  return newValidates;
};

export const connect = (validators, Component) => {
  const displayName = Component.displayName || Component.name || 'Component';

  class ValidComponent extends React.Component {
    constructor(props) {
      super(props);
      this.state = {
        valid: true,
        message: '',
        validatorKeys: [],
      };

      this.validate = this.validate.bind(this);
    }

    componentWillMount() {
      const validatorKeys = Object.keys(validators).filter(name =>
        Object.prototype.hasOwnProperty.call(this.props, name),
      );
      this.setState({
        validatorKeys,
      });
    }

    componentWillReceiveProps(nextProps) {
      const validatorKeys = Object.keys(validators).filter(name =>
        Object.prototype.hasOwnProperty.call(nextProps, name),
      );
      this.setState({
        validatorKeys,
      });
    }

    validate(value) {
      const { onInvalid = () => {}, onValid = () => {} } = this.props;
      const { validatorKeys } = this.state;
      let message = '';

      const status = validatorKeys.every(name => {
        if (!validators[name].method(value, this.props[name])) {
          if (typeof validators[name].message === 'string') {
            // Display normal string message
            message = validators[name].message;
          } else {
            // Use template to generate message
            message = validators[name].message(value, this.props[name]);
          }
          return false;
        }
        return true;
      });

      this.setState(
        {
          valid: status,
          message,
        },
        () => {
          if (!status) {
            onInvalid(value);
          } else {
            onValid(value);
          }
        },
      );
    }

    render() {
      const childrenProps = {};
      const { validatorKeys, valid, message } = this.state;

      Object.keys(this.props).forEach(name => {
        if (validatorKeys.indexOf(name) === -1 && name !== 'onValid' && name !== 'onInvalid') {
          childrenProps[name] = this.props[name];
        }
      });

      return (
        <Component {...childrenProps} valid={valid} message={message} validate={this.validate} />
      );
    }
  }
  ValidComponent.displayName = `Valle(${displayName})`;

  return ValidComponent;
};

export const template = (strings, ...keys) => (validateValue, props) => {
  const result = [strings[0]];
  keys.forEach((key, i) => {
    let value = '';
    if (key === 0) {
      value = validateValue;
    } else if (key === 1) {
      value = typeof props === 'string' || !isNaN(props) || props instanceof RegExp
        ? props
        : props[key];
    } else {
      value = props[key];
    }
    result.push(value, strings[i + 1]);
  });
  return result.join('');
};
