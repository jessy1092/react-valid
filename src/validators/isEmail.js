import isEmail from 'validator/lib/isEmail';

import { isEmpty } from './util';

export default {
  isEmail: {
    method: value => isEmpty(value) || isEmail(value),
    message: 'This is not a valid email address. Please check your email and try again.',
  },
};
