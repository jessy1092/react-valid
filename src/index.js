import { addMethod, addMethods, connect, template } from './core';

import isEmail from './validators/isEmail';
import contains from './validators/contains';
import equals from './validators/equals';
import isBoolean from './validators/isBoolean';
import isFloat from './validators/isFloat';
import isISO8601 from './validators/isISO8601';
import isInt from './validators/isInt';

export function createValle() {
  let validators = {};

  return {
    addMethod(name, method, message) {
      validators = addMethod(validators, name, method, message);
    },
    addMethods(appendValidators) {
      validators = addMethods(validators, appendValidators);
    },
    connect(Component) {
      return connect(validators, Component);
    },
    template,
  };
}

const valle = createValle();

valle.addMethods(isEmail);
valle.addMethods(contains);
valle.addMethods(equals);
valle.addMethods(isBoolean);
valle.addMethods(isFloat);
valle.addMethods(isISO8601);
valle.addMethods(isInt);

export default valle;
