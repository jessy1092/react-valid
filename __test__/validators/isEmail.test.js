import React from 'react';
import { mount } from 'enzyme';

import { createValle } from '../../src';
import isEmail from '../../src/validators/isEmail';

import Input from '../fakeComponent/Input';

test('It sould be invalid if input without email', () => {
  const valle = createValle();
  valle.addMethods(isEmail);

  const ValidInput = valle.connect(Input);
  const wrapper = mount(<ValidInput isEmail />);

  wrapper.find('input').simulate('change', {
    target: {
      value: 'jessy1092',
    },
  });

  expect(wrapper.state('valid')).toEqual(false);
  expect(wrapper.state('message')).toEqual(
    'This is not a valid email address. Please check your email and try again.',
  );
  expect(wrapper).toMatchSnapshot();
});

test('It sould be valid if input email', () => {
  const valle = createValle();
  valle.addMethods(isEmail);

  const ValidInput = valle.connect(Input);
  const wrapper = mount(<ValidInput isEmail />);

  wrapper.find('input').simulate('change', {
    target: {
      value: 'jessy1092@gmail.com',
    },
  });

  expect(wrapper.state('valid')).toEqual(true);
  expect(wrapper.state('message')).toEqual('');
  expect(wrapper).toMatchSnapshot();
});

test('It sould be valid if input is empty', () => {
  const valle = createValle();
  valle.addMethods(isEmail);

  const ValidInput = valle.connect(Input);
  const wrapper = mount(<ValidInput isEmail />);

  wrapper.find('input').simulate('change', {
    target: {
      value: '',
    },
  });

  expect(wrapper.state('valid')).toEqual(true);
  expect(wrapper.state('message')).toEqual('');
  expect(wrapper).toMatchSnapshot();
});
