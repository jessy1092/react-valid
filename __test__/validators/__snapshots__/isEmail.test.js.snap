// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It sould be invalid if input without email 1`] = `
<Valle(Input)
  isEmail={true}
>
  <Input
    message="This is not a valid email address. Please check your email and try again."
    valid={false}
    validate={[Function]}
  >
    <div>
      <input
        onChange={[Function]}
      />
      <div>
        This is not a valid email address. Please check your email and try again.
      </div>
    </div>
  </Input>
</Valle(Input)>
`;

exports[`It sould be valid if input email 1`] = `
<Valle(Input)
  isEmail={true}
>
  <Input
    message=""
    valid={true}
    validate={[Function]}
  >
    <div>
      <input
        onChange={[Function]}
      />
      <div />
    </div>
  </Input>
</Valle(Input)>
`;
